agent/src/main/java/com/code_intelligence/jazzer/runtime/CoverageMap.java

```diff
@@ -21,18 +21,56 @@
 
 /**
  * Represents the Java view on a libFuzzer 8 bit counter coverage map.
- * By using a direct ByteBuffer, the counter array is shared directly with
- * native code.
+ * By using a direct ByteBuffer, the counters are shared directly with native
+ * code.
  */
 final public class CoverageMap {
-  public static ByteBuffer counters = ByteBuffer.allocateDirect(0);
+  private static final int MAX_NUM_COUNTERS = 1 << 20;
+
+  /**
+   * The collection of coverage counters directly interacted with by classes
+   * that are instrumented for coverage.
+   * The instrumentation assumes that this is always one contiguous block of
+   * memory, so it is allocated once at maximum size. Using a larger number here
+   * increases the memory usage of all fuzz targets, but has otherwise no impact
+   * on performance.
+   */
+  public static final ByteBuffer counters = ByteBuffer.allocateDirect(MAX_NUM_COUNTERS);
+
+  static {
+    initialize(counters);
+  }
+
+  private static final int INITIAL_NUM_COUNTERS = 1 << 9;
+
+  static {
+    registerNewCounters(0, INITIAL_NUM_COUNTERS);
+  }
+
+  /**
+   * The number of coverage counters that are currently registered with
+   * libFuzzer. This number grows dynamically as classes are instrumented and
+   * should be kept as low as possible as libFuzzer has to iterate over the
+   * whole map for every execution.
+   */
+  private static int currentNumCounters = INITIAL_NUM_COUNTERS;
 
   // Called via reflection.
   @SuppressWarnings("unused")
   public static void enlargeIfNeeded(int nextId) {
-    if (nextId >= counters.capacity()) {
-      registerNewCoverageCounters();
-      System.out.println("INFO: New number of inline 8-bit counters: " + counters.capacity());
+    int newNumCounters = currentNumCounters;
+    while (nextId >= newNumCounters) {
+      newNumCounters = 2 * currentNumCounters;
+      if (newNumCounters > MAX_NUM_COUNTERS) {
+        System.out.printf(
+            "ERROR: Maximum number (%s) of coverage counters exceeded%n", MAX_NUM_COUNTERS);
+        System.exit(1);
```

Comment:
I think that this should always be an error, not a warning. Fuzzers print enough log lines that warnings could get lost and this error condition would seriously undermine the fuzzer performance. Since most of the classes will be discovered closer to startup, this error would usually be thrown early in the fuzzing run.

I made the error more expressive and added an escape hatch for users to increase the limit, although I doubt that would ever be the right thing to do (rather than break up the fuzz target).

Reply:
You are right about the error. I find the error message now very informative and it recommends to the user to limit the scope of the fuzz test.